flink-libraries/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/DynamicResult.java

```diff
@@ -0,0 +1,253 @@
+				return handleMissingResult();
+			}
+		}
+	}
+
+	public TypedResult<Integer> snapshot(int pageSize) {
+		synchronized (resultLock) {
+			// retrieval thread is dead and there are no results anymore
+			// or program failed
+			if ((!retrievalThread.isRunning && isLastSnapshot) || executionException != null) {
+				return handleMissingResult();
+			}
+			// this snapshot is the last result that can be delivered
+			else if (!retrievalThread.isRunning) {
+				isLastSnapshot = true;
+			}
+
+			this.pageSize = pageSize;
+			snapshot.clear();
+			snapshot.addAll(materializedTable);
+
+			// at least one page
+			final int pageCount = Math.max(1, (int) Math.ceil(((double) snapshot.size() / pageSize)));
+			return TypedResult.payload(pageCount);
+		}
+	}
+
+	public List<Row> retrievePage(int page) {
+		final int pageCount = Math.max(1, (int) Math.ceil(((double) snapshot.size()) / pageSize));
+		if (page <= 0 || page > pageCount) {
+			throw new SqlExecutionException("Invalid page '" + page + "'.");
+		}
+
+		return snapshot.subList(pageSize * (page - 1), Math.min(snapshot.size(), pageSize * page));
+	}
+
+	public TableSink<?> getTableSink() {
+		return collectTableSink;
+	}
+
+	public void close() {
+		retrievalThread.isRunning = false;
+		retrievalThread.interrupt();
+		monitoringThread.interrupt();
+		collectTableSink.close();
+	}
+
+	// --------------------------------------------------------------------------------------------
+
+	private <T> TypedResult<T> handleMissingResult() {
+		// check if the monitoring thread is still there
+		// we need to wait until we know what is going on
+		if (monitoringThread.isAlive()) {
+			return TypedResult.empty();
+		}
+		// the job finished with an exception
+		else if (executionException != null) {
+			throw executionException;
+		}
+		// we assume that a bounded job finished
+		else {
+			return TypedResult.endOfStream();
+		}
+	}
+
+	// --------------------------------------------------------------------------------------------
+
+	private class JobMonitoringThread extends Thread {
+
+		@Override
+		public void run() {
+			try {
+				program.run();
+			} catch (SqlExecutionException e) {
+				executionException = e;
+			}
+		}
+	}
+
+	// --------------------------------------------------------------------------------------------
+
+	private class ResultRetrievalThread extends Thread {
+
+		public volatile boolean isRunning = true;
+
+		@Override
+		public void run() {
+			try {
+				while (isRunning && collectTableSink.hasNext()) {
+					final Tuple2<Boolean, Row> change = collectTableSink.next();
+
+					if (isChangelog) {
+						processForChangelog(change);
+					} else {
+						processForTable(change);
+					}
+				}
+			} catch (RuntimeException e) {
+				// ignore socket exceptions
+			}
+
+			// no result anymore
+			// either the job is done or an error occurred
+			isRunning = false;
+		}
+
+		private void processForChangelog(Tuple2<Boolean, Row> change) {
+			synchronized (resultLock) {
```

Comment:
I think this will block the job and eventually lead to back pressure. Why not collecting change rows in a list like a table and returning them as a batch? This should reduce the bottleneck effect a bit.